File: backend/src/main/java/github/lms/lemuel/xr/values/application/GetValueProfileUseCase.java
package github.lms.lemuel.xr.values.application;

import github.lms.lemuel.xr.values.adapter.out.persistence.UserValuePracticeJpaEntity;
import github.lms.lemuel.xr.values.adapter.out.persistence.UserValuePracticeRepository;
import github.lms.lemuel.xr.values.adapter.out.persistence.UserValueProfileJpaEntity;
import github.lms.lemuel.xr.values.adapter.out.persistence.UserValueProfileRepository;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/** 사용자 7 가치 프로파일 + 최근 7일 실천 통계 + CDR Index. */
@Service
@RequiredArgsConstructor
public class GetValueProfileUseCase {

    private final UserValueProfileRepository profiles;
    private final UserValuePracticeRepository practices;

    @Transactional(readOnly = true)
    public Result execute(UUID userId) {
        UserValueProfileJpaEntity p = profiles.findByUserId(userId).orElse(null);
        Map<String, Object> valuesJson = p == null ? new HashMap<>() : p.getValuesJson();

        OffsetDateTime sevenDaysAgo = OffsetDateTime.now(ZoneOffset.UTC).minusDays(7);
        List<UserValuePracticeJpaEntity> recent = practices.findRecent(userId, sevenDaysAgo);

        // value_id (1~7) 별 카운트
        Map<String, Integer> countByValue = new HashMap<>();
        for (UserValuePracticeJpaEntity r : recent) {
            countByValue.merge(String.valueOf(r.getValueId()), 1, Integer::sum);
        }

        int totalPractices = recent.size();
        int cdrIndex = computeCdrIndex(totalPractices);
        String tier = tierOf(cdrIndex);

        return new Result(valuesJson, totalPractices, countByValue, cdrIndex, tier);
    }

    /**
     * Civil Defense Readiness Index — CROSS-MAPPING-VR-AR.md §6.
     *
     * <p>49 = 7 가치 × 7일. 100 = 매일 7 가치 모두 실천. 0 = 7일간 실천 0건.</p>
     */
    static int computeCdrIndex(int totalPractices) {
        int maxPossible = 49;
        int raw = Math.min(maxPossible, totalPractices) * 100 / maxPossible;
        return Math.max(0, Math.min(100, raw));
    }

    static String tierOf(int cdr) {
        if (cdr >= 80) return "VERY_READY";
        if (cdr >= 50) return "NORMAL";
        return "BUILD_UP";
    }

    public record Result(
            Map<String, Object> valuesJson,
            int totalPractices7d,
            Map<String, Integer> countByValue,
            int cdrIndex,
            String tier
    ) {}
}

File: backend/src/main/kotlin/github/lms/lemuel/xr/values/application/GetValueProfileUseCase.kt
package github.lms.lemuel.xr.values.application

import github.lms.lemuel.xr.values.application.port.out.UserValuePracticePort
import github.lms.lemuel.xr.values.application.port.out.UserValueProfilePort
import github.lms.lemuel.xr.values.domain.CdrIndex
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.UUID

/** 사용자 7 가치 프로파일 + 최근 7일 실천 통계 + CDR Index. */
@Service
class GetValueProfileUseCase(
    private val profiles: UserValueProfilePort,
    private val practices: UserValuePracticePort,
) {

    @Transactional(readOnly = true)
    fun execute(userId: UUID): Result {
        val profile = profiles.findByUserId(userId).orElse(null)
        val valuesJson: Map<String, Any?> = profile?.valuesJson ?: emptyMap()

        val sevenDaysAgo = OffsetDateTime.now(ZoneOffset.UTC).minusDays(7)
        val recent = practices.findRecent(userId, sevenDaysAgo)

        // value_id (1~7) 별 카운트
        val countByValue = HashMap<String, Int>()
        for (r in recent) {
            countByValue.merge(r.valueId.toString(), 1, Int::plus)
        }

        val totalPractices = recent.size
        val cdr = CdrIndex.fromPractices(totalPractices)

        return Result(valuesJson, totalPractices, countByValue, cdr.value, cdr.tier)
    }

    data class Result(
        val valuesJson: Map<String, Any?>,
        val totalPractices7d: Int,
        val countByValue: Map<String, Int>,
        val cdrIndex: Int,
        val tier: String,
    )
}
